src/Microsoft.AspNetCore.SystemWebAdapters/Adapters/ProxyHeaderModule.Framework.cs

```diff
@@ -75,4 +118,26 @@ private static void UseForwardedFor(HttpRequest request)
             request.ServerVariables.Set("REMOTE_HOST", remote);
         }
     }
+
+    private static void StoreOriginalServerVariables(HttpRequest request)
+    {
+        request.ServerVariables.Set("ORIGINAL_SERVER_NAME", request.ServerVariables["SERVER_NAME"]);
+        request.ServerVariables.Set("ORIGINAL_SERVER_PORT", request.ServerVariables["SERVER_PORT"]);
+        request.ServerVariables.Set("ORIGINAL_SERVER_PROTOCOL", request.ServerVariables["SERVER_PROTOCOL"]);
+    }
+
+    private static int? GetPort(string? portString, string protocol)
+    {
+        if (int.TryParse(portString, out var port))
+        {
+            return port;
+        }
+
+        return protocol.ToUpperInvariant() switch
```

Comment:
nit: this would do a string allocation every request

Reply:
Good point. I'll just do a couple invariant comparisons instead.